app/models/concerns/orchestration/dns.rb

```diff
@@ -1,175 +1,211 @@
+    return unless dns6? || @dns_aaaa_record
+    return unless ip6_available?
+    @dns_aaaa_record ||= Net::DNS::AAAARecord.new dns_aaaa_record_attrs
   end
 
   def dns_ptr_record
-    return unless reverse_dns? or @dns_ptr_record
+    return unless reverse_dns? || @dns_ptr_record
     @dns_ptr_record ||= Net::DNS::PTR4Record.new reverse_dns_record_attrs
   end
 
-  def del_dns_a_record_safe
-    if dns_a_record
-      begin
-        del_dns_a_record
-      rescue => e
-        Foreman::Logging.exception "Proxy failed to delete DNS a_record for #{name}, #{ip}", e, :level => :error
-      end
-    end
+  def dns_ptr6_record
+    return unless reverse_dns6? || @dns_ptr6_record
+    @dns_ptr6_record ||= Net::DNS::PTR6Record.new reverse_dns6_record_attrs
   end
 
-  def del_dns_ptr_record_safe
-    if dns_ptr_record
-      begin
-        del_dns_ptr_record
-      rescue => e
-        Foreman::Logging.exception "Proxy failed to delete DNS ptr_record for #{name}, #{ip}", e, :level => :error
+  RECORD_TYPES.each do |record_type|
+    define_method("del_dns_#{record_type}_record_safe") do # def del_dns_a_record_safe
+      if send(:"dns_#{record_type}_record") # dns_a_record
+        begin
+          send(:"del_dns_#{record_type}_record") # del_dns_a_record
+        rescue => e
+          Foreman::Logging.exception "Proxy failed to delete DNS #{record_type}_record for #{name}, #{ip}, #{ip6}", e, :level => :error
+        end
       end
     end
   end
 
   protected
 
-  def recreate_a_record
-    set_dns_a_record unless dns_a_record.nil? || dns_a_record.valid?
-  end
-
-  def recreate_ptr_record
-    set_dns_ptr_record unless dns_ptr_record.nil? || dns_ptr_record.valid?
-  end
+  RECORD_TYPES.each do |record_type|
+    record = :"dns_#{record_type}_record"
 
-  def set_dns_a_record
-    dns_a_record.create
-  end
-
-  def set_conflicting_dns_a_record
-    dns_a_record.conflicts.each { |c| c.create }
-  end
-
-  def set_dns_ptr_record
-    dns_ptr_record.create
-  end
-
-  def set_conflicting_dns_ptr_record
-    dns_ptr_record.conflicts.each { |c| c.create }
-  end
+    define_method("recreate_#{record_type}_record") do # def recreate_a_record
+      # set_dns_a_record unless dns_a_record.nil? || dns_a_record.valid?
+      send(:"set_dns_#{record_type}_record") unless send(record).nil? || send(record).valid?
+    end
 
-  def del_dns_a_record
-    dns_a_record.destroy
-  end
+    define_method("set_dns_#{record_type}_record") do # def set_dns_a_record
+      send(record).create # dns_a_record.create
+    end
 
-  def del_conflicting_dns_a_record
-    dns_a_record.conflicts.each { |c| c.destroy }
-  end
+    define_method("set_conflicting_dns_#{record_type}_record") do # def set_conflicting_dns_a_record
+      send(record).conflicts.each { |c| c.create } # dns_a_record.conflicts.each { |c| c.create }
+    end
 
-  def del_dns_ptr_record
-    dns_ptr_record.destroy
-  end
+    define_method("del_dns_#{record_type}_record") do # def del_dns_a_record
```

Comment:
I wonder if it's worth splitting this class now somehow.. gotta pick up [Refactoring](https://www.amazon.com/Refactoring-Improving-Design-Existing-Code/dp/0201485672) again...